File: DSA/src/Questions/StackQueue/StackUsingQueue.java
package Questions.StackQueue;

import java.util.LinkedList;
import java.util.Queue;

/*
 https://leetcode.com/problems/implement-stack-using-queues/description/
 Time Complexity:
Push: O(1) for adding an element to the stack.
Pop: O(1) for removing the top element from the stack.
Top: O(1) for getting the top element from the stack.
IsEmpty: O(1) for checking if the stack is empty.
IsFull: O(1) for checking if the stack is full.
Space Complexity: O(n) for storing the elements in the array, where n is the capacity of the stack.
 */
public class StackUsingQueue {
    Queue<Integer> q1 = new LinkedList<>();
    Queue<Integer> q2 = new LinkedList<>();

    public StackUsingQueue() {

    }

    public void push(int x) {
        while (q1.size() != 0) {
            q2.add(q1.poll());
        }
        q1.add(x);
        while (q2.size() != 0) {
            q1.add(q2.poll());
        }
    }

    public int pop() {
        return q1.poll();
    }

    public int top() {
        return q1.peek();
    }

    public boolean empty() {
        return q1.size() == 0;
    }

    public static void main(String[] args) {
        StackUsingQueue stack = new StackUsingQueue();
        stack.push(1);
        stack.push(2);
        stack.push(3);
        stack.push(4);
        stack.push(5);
        System.out.println("Top element of the stack: " + stack.top());
    }
}

File: DSA/Python/Questions/StackQueue/StackUsingQueue.py
"""
# Question: Implement Stack using Queues
# Link: https://leetcode.com/problems/implement-stack-using-queues/

# Implement stack operations using queues

# Time Complexity:
# - Push: O(1)
# - Pop: O(1)
# - Top: O(1)
# - Empty: O(1)

# Space Complexity: O(n)

# Algorithm:
# 1. Use two queues for stack simulation
# 2. Transfer elements between queues for operations
# 3. Maintain LIFO order
# 4. Optimize push operation

# Key Components:
# - MyStack class implementation
# - Two queues for storage
# - Element transfer logic
"""

from collections import deque


class MyStack:
    def __init__(self):
        self.q1 = deque()
        self.q2 = deque()

    def push(self, x: int) -> None:
        while self.q1:
            self.q2.append(self.q1.popleft())
        self.q1.append(x)
        while self.q2:
            self.q1.append(self.q2.popleft())

    def pop(self) -> int:
        return self.q1.popleft()

    def top(self) -> int:
        return self.q1[0]

    def empty(self) -> bool:
        return len(self.q1) == 0


def main():
    stack = MyStack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    print(f"Top element of the stack: {stack.top()}")


if __name__ == "__main__":
    main()
